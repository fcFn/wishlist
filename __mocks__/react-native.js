const RN = require('react-native');

RN.AppState.removeEventListener = function() {};
RN.BackAndroid.removeEventListener = jest.fn();

module.exports = RN;
