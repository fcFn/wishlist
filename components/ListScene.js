// The most important component which contains the list, the data model
// for the list to display, and some of the controls.

import React, {Component} from 'react';
import {
  Text,
  View,
  TouchableNativeFeedback,
  ListView,
  ToastAndroid,
  Alert,
  Platform,
  Keyboard,
  AppState,
} from 'react-native';

import AddEditItemPanel from './AddEditItemPanel';
import ItemList from './ItemList';
import HelpModal from './HelpModal';
import MenuModal from './MenuModal';
import {packItems, unpackItems} from '../helpers/qr';
import RNFetchBlob from 'react-native-fetch-blob';
import qr from '../helpers/rn-qr-image';
import I18n from 'react-native-i18n';

import styles from '../configs/styles';

export default class ListScene extends Component {
  constructor() {
    super();
    this.EMPTY_ITEM_TEXT = I18n.t('emptyList');
    if (__DEV__) {
      this.defaultItems = require('./TestItems');
    } else {
      this.defaultItems = [
        {
          name: this.EMPTY_ITEM_TEXT,
          done: false,
        },
      ];
    }
    this.ds = new ListView.DataSource({rowHasChanged: (r1, r2) => r1 !== r2});
    this.state = {
      items: [{}],
      inputActive: false,
      offsetForKeyboard: 0,
      itemToEdit: null,
      generatingQR: false,
      listChanged: false,
      menuModalVisible: false,
      helpModalVisible: false,
      helpModalType: 'notFirst',
    };
  };

  componentWillMount() {
    this._keyboardDidHide = this._keyboardDidHide.bind(this);
    this.handleAppStateChange = this.handleAppStateChange.bind(this);
    Keyboard.addListener('keyboardDidHide', this._keyboardDidHide);
    AppState.addEventListener('change', this.handleAppStateChange);
    this.checkForFirstRun();
  }

  componentWillUnmount() {
    this.saveList('onExit');
    Keyboard.removeListener('keyboardDidHide', this._keyboardDidHide);
    AppState.removeEventListener('change', this.handleAppStateChange);
  }

  _keyboardDidHide() {
    // Hiding keyboard should exit input mode
    // Might be inconvenient for external kb usecases
    this.setState({inputActive: false});
  }

  handleAppStateChange() {
    if ((AppState.currentState === 'background' ||
        AppState.currentState === 'inactive') &&
        this.userAddedItem) {
      this.saveList('onExit');
    }
  }

  removeItem(item) {
    let items = this.copyItems();
    let indexToRemoveAt = items.findIndex((element) => element.name === item);
    items.splice(indexToRemoveAt, 1);
    if (!items.length) {
      items = [{name: this.EMPTY_ITEM_TEXT, done: false}];
      this.userAddedItem = false;
    }
    this.setState({items: items});
    this.listChanged = true;
  }

  checkItem(item) {
    // It also unchecks checked item
    let items = this.copyItems();
    let indexofItemToCheck = items.findIndex(
        (element) => element.name === item
      );
    items[indexofItemToCheck] = {
      name: item,
      done: !items[indexofItemToCheck].done,
    };
    this.setState({items: items});
    this.listChanged = true;
  }

  copyItems() {
    return JSON.parse(JSON.stringify(this.state.items));
  }

  toggleInput() {
    this.setState({inputActive: !this.state.inputActive});
  }

  itemAlreadyExists(itemName, itemIndex) {
    if (Platform.OS === 'android') {
      ToastAndroid.show((I18n.t('itemAlreadyExists'))
                        .replace('${itemName}', itemName), 2);
    }
    // Scroll all the way up and then down to the related item
    this.itemList.listView.scrollTo({
      x: 0,
      y: -(this.ds.getRowCount() * 45),
      animate: false});
    this.itemList.listView.scrollTo({
      x: 0,
      y: itemIndex * 45,
      animate: false});
    let items = this.copyItems();
    items[itemIndex].alreadyExistsFlag = true;
    this.setState({items: items});
    setTimeout(() => {
      let items = this.copyItems();
      items[itemIndex].alreadyExistsFlag = false;
      this.setState({items: items});
    }, 3000);
  }

  addButtonLongPress() {
    this.setState({menuModalVisible: true});
  }

  _capitalize(string) {
    return string[0].toUpperCase() + string.slice(1);
  }

  addNewItem(event) {
    let itemName = event.nativeEvent.text;
    if (itemName === '') {
      this.toggleInput();
      return;
    }
    // autoCapitalize prop doesn't work on some devices
    itemName = this._capitalize(itemName);
    let itemIndex = this.state.items.findIndex(
      (element) => element.name === itemName);
    if (itemIndex !== -1) {
      this.itemAlreadyExists(itemName, itemIndex);
    } else {
      let items;
      if (!this.userAddedItem) {
        items = [];
      } else {
        items = this.copyItems();
      }
      this.userAddedItem = true;
      items.push({name: itemName, done: false});
      this.setState({items: items});
      this.listChanged = true;
    }
    this.toggleInput();
  }

  editItem(event) {
    if (event.nativeEvent.text !== '') {
    // autoCapitalize prop doesn't work on some devices
      let capitalizedText = this._capitalize(event.nativeEvent.text);
      if (capitalizedText !== this.state.itemToEdit.name) {
        let items = this.copyItems();
        let itemToEditIndex = items.findIndex((element) =>
        element.name === this.state.itemToEdit.name);
        let existingItemIndex = items.
                  findIndex((element) => element.name === capitalizedText);
        if (existingItemIndex !== -1) {
          this.itemAlreadyExists(items[existingItemIndex].name,
          existingItemIndex);
        } else {
          items[itemToEditIndex].name = capitalizedText;
          this.setState({items: items});
          this.listChanged = true;
          this.userAddedItem = true;
        }
      }
    }
    this.props.navigator.pop();
  }

  recordItemToEdit(item) {
    // An item calls this when it is long pressed
    let itemToEdit = this.state.items.find(
      (element) => element.name === item);
    this.setState({itemToEdit: itemToEdit});
    this.props.navigator.push({
      title: 'Edit item',
      index: this.props.route.index + 1,
    });
  }

  checkForFirstRun() {
    RNFetchBlob.fs.exists(RNFetchBlob.fs.dirs.DocumentDir + '/config')
    .then((exists) => {
      if (!exists) {
        RNFetchBlob.fs.writeFile(RNFetchBlob.fs.dirs.DocumentDir + '/config',
          JSON.stringify({lastRunTime: Date.now()}), 'utf8')
        .then(() =>
          this.setState({helpModalVisible: true, helpModalType: 'first'})
        );
      } else {
        RNFetchBlob.fs
          .readFile(RNFetchBlob.fs.dirs.DocumentDir + '/config', 'utf8')
        .then((data) => {
          let config = JSON.parse(data);
          if (Date.now() - config.lastRunTime > 2592e6 ) {
            console.log(Date.now());
            this.setState({helpModalVisible: true, helpModalType: 'notFirst'});
          }
        })
        .then(() =>
        RNFetchBlob.fs.writeFile(RNFetchBlob.fs.dirs.DocumentDir + '/config',
          JSON.stringify({lastRunTime: Date.now()}), 'utf8'));
      }
    });
  }

  showHelpModal() {
    this.closeMenuModal();
    this.setState({helpModalVisible: true, helpModalType: 'onDemand'});
  }

  readQR() {
    this.props.navigator.push({
      title: 'Read QRCode',
      index: this.props.route.index + 1,
    });
  }

  processQR(data) {
    this.props.navigator.pop();
    if (data.type === 'QR_CODE') {
      try {
        this.setState({items: unpackItems(data.data)});
        this.listChanged = true;
      } catch (e) {
        if (Platform.OS === 'android') {
          ToastAndroid.show(I18n.t('invalidQR'), 2);
        }
      }
    }
  }

  generateQR() {
    this.closeMenuModal();
    this.setState({generatingQR: true});
    let image;
    try {
      image = qr.imageSync(packItems(this.state.items),
      {ec_level: 'L', margin: 0});
    } catch (e) {
      if (e = 'Too much data') {
        alert(I18n.t('qrListTooLarge'));
        return;
      } else {
        throw (e);
      }
    } finally {
      this.setState({generatingQR: false});
    }
    return RNFetchBlob
     .fs
     .writeFile(RNFetchBlob.fs.dirs.DocumentDir +
                '/qr.png', Array.from(image), 'ascii')
     .then(() => {
       this.setState({generatingQR: false});
       this.props.navigator.push({
         title: 'Show QRCode',
         index: this.props.route.index + 1,
       });
     });
  }


  saveList(type) {
    this.closeMenuModal();
    let filename = type === 'onExit' ? '/onExit.list' : '/1.list';
    let items = this.copyItems();
    // strip non-informative flags
    for (let item of items) {
      {
        delete item.alreadyExistsFlag;
      }
    }
    return RNFetchBlob.fs.writeFile(
      RNFetchBlob.fs.dirs.DocumentDir + filename,
      JSON.stringify(items),
      'utf8').then(() => {
        if (Platform.OS === 'android' && type !== 'onExit') {
          ToastAndroid.show(I18n.t('listSaved'), 2);
        }
        this.listChanged = false;
      });
  }

  loadList(type) {
    this.closeMenuModal();
    let filename = type === 'onStart' ? '/onExit.list' : '/1.list';
    return RNFetchBlob.fs.readFile(
      RNFetchBlob.fs.dirs.DocumentDir + filename,
      'utf8'
    )
    .then(
      (file) => {
        let items = JSON.parse(file);
        if (Platform.OS === 'android' && type !== 'onStart') {
          ToastAndroid.show(I18n.t('listLoaded'), 2);
        }
        this.userAddedItem = true;
        this.setState({items: items});
        this.listChanged = false;
      },
      () => {
        if (Platform.OS === 'android') {
          ToastAndroid.show(I18n.t('noListFound'), 2);
        }
      });
  }

  newList() {
    this.setState({items: [
      {
        name: this.EMPTY_ITEM_TEXT,
        done: false,
      },
    ]}
    );
    this.userAddedItem = false;
    this.listChanged = true;
  }

  newListAlert() {
    this.closeMenuModal();
    if(this.listChanged && this.userAddedItem) {
      Alert.alert(
      I18n.t('createNewListTitle'),
      I18n.t('createNewListMsg'),
        [
        {text: 'OK', onPress: this.newList.bind(this)},
        {text: 'Cancel'},
        ]
    );
    } else {
      this.newList();
    }
  }

  loadListAlert() {
    this.closeMenuModal();
    if (this.listChanged && this.userAddedItem) {
      Alert.alert(
      I18n.t('loadListTitle'),
      I18n.t('loadListMsg'),
        [
        {text: 'OK', onPress: this.loadList.bind(this)},
        {text: 'Cancel'},
        ]
    );
    } else {
      this.loadList();
    }
  }

  readQRAlert() {
    this.closeMenuModal();
    if (this.listChanged && this.userAddedItem) {
      Alert.alert(
        I18n.t('readQRAlertTitle'),
        I18n.t('readQRAlertMsg'),
        [
          {text: 'OK', onPress: this.readQR.bind(this)},
          {text: 'Cancel'},
        ]
      );
    } else {
      this.readQR();
    }
  }

  componentDidMount() {
    return RNFetchBlob.fs.exists(
      RNFetchBlob.fs.dirs.DocumentDir + '/onExit.list')
      .then((exists) => {
        if (exists) {
          this.loadList('onStart');
        } else {
          this.setState({items: this.defaultItems});
        }
      }
    );
  }

  closeHelpModal() {
    this.setState({helpModalVisible: false});
  }

  closeMenuModal() {
    this.setState({menuModalVisible: false});
  }

  render() {
    let newItemPanel = this.state.inputActive ?
                       <AddEditItemPanel style={styles.newItemInput}
                       textInputStyle={styles.textInputStyleAddEditItem}
                       onSubmitEditing={this.addNewItem.bind(this)} /> :
                       null;
    let addButton = this.state.inputActive ? null : <TouchableNativeFeedback
      onPress={this.toggleInput.bind(this)}
      onLongPress={this.addButtonLongPress.bind(this)}
      >
      <View style={styles.addButton}>
        <Text style={styles.addButtonText}>
          ⊕
        </Text>
      </View>
    </TouchableNativeFeedback>;
    return (
      <View style={styles.container}>
        <HelpModal
          type={this.state.helpModalType}
          visible={this.state.helpModalVisible}
          closeHelpModal={this.closeHelpModal.bind(this)} />
        <MenuModal
          visible={this.state.menuModalVisible}
          closeMenuModal={this.closeMenuModal.bind(this)}
          generatingQR={this.state.generatingQR}
          generateQR={this.generateQR.bind(this)}
          readQRAlert={this.readQRAlert.bind(this)}
          newListAlert={this.newListAlert.bind(this)}
          loadListAlert={this.loadListAlert.bind(this)}
          saveList={this.saveList.bind(this)}
          showHelpModal={this.showHelpModal.bind(this)}
        />
        <ItemList
        ref={(itemList) => this.itemList = itemList}
        inputActive={this.state.inputActive}
        dataSource={this.ds.cloneWithRows(this.state.items)}
        checkItem = {this.checkItem.bind(this)}
        removeItem = {this.removeItem.bind(this)}
        recordItemToEdit={this.recordItemToEdit.bind(this)}
        initialSize={this.state.items.length}
        />
        {newItemPanel}
        <View style={styles.buttons}>
        {addButton}
      </View>
    </View>
    );
  };
}
