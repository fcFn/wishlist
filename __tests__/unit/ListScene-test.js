import {Alert} from 'react-native';
import qr from '../../helpers/qr';
import React from 'react';
import i18n from 'react-native-i18n';
import '../../configs/translations';
import ListScene from '../../components/ListScene';
import _ from 'lodash';
import renderer from 'react-test-renderer';
import defaultItems from '../../components/TestItems';

jest.mock('react-native-fetch-blob');
jest.mock('../../helpers/qr');

let testItems = [
  {
    name: 'Apples',
    done: false,
  }, {
    name: 'Oranges',
    done: true,
  },
];

// setState mock for the component
ListScene.prototype.setState = function(state) {
  _.extend(this.state, state);
};

let ls;
let RNFetchBlob = require('react-native-fetch-blob');
let mockPush = jest.fn();
let mockPop = jest.fn();

// Various mocks
beforeEach(() => {
  ls = new ListScene;
  ls.spyCloseMenuModal = jest.spyOn(ls, 'closeMenuModal');
  ls.props = {
    navigator: {
      push: mockPush,
      pop: mockPop,
    },
    route: {
      index: 0,
    },
  };
  _.set(ls, 'itemList.listView.scrollTo', jest.fn());
  ls.spyItemAlreadyExists = jest.spyOn(ls, 'itemAlreadyExists');
  RNFetchBlob.fs.files = {};
});

describe('ListScene', () => {
  describe('component', () => {
    it('renders correctly', () => {
      const tree = renderer.create(<ListScene/>);
      expect(tree).toMatchSnapshot();
    });

    it('loads default items', () => ls.componentDidMount().
      then(() => expect(ls.state.items).toEqual(defaultItems)));

    it('saves list', () => {
      ls.state.items = _.cloneDeep(testItems);
      return ls
        .saveList()
        .then(() => {
          let data = RNFetchBlob.fs.files['/1.list'];
          expect(JSON.parse(data)).toEqual(testItems);
        });
    });

    it('loads list', () => {
      RNFetchBlob.fs.files['/1.list'] = JSON.stringify(testItems);
      return ls
        .loadList()
        .then(() => expect(ls.state.items).toEqual(testItems));
    });

    it('loads list after rendering', () => {
      RNFetchBlob.fs.files['/onExit.list'] = JSON.stringify(testItems);
      return ls
        .componentDidMount()
        .then(() => expect(ls.state.items).toEqual(testItems));
    });

    it('shows the help modal', () => {
      ls.showHelpModal();
      expect(ls.spyCloseMenuModal).toBeCalled();
      expect(ls.state.helpModalVisible).toBe(true);
      expect(ls.state.helpModalType).toBe('onDemand');
    });

    it('switches to QR reading scene', () => {
      ls.readQR();
      expect(ls.props.navigator.push).toBeCalledWith(
        {title: 'Read QRCode',
          index: ls.props.route.index + 1}
      );
    });

    describe('addNewItem', () => {
      beforeEach(() => {
        ls.state.items = _.cloneDeep(testItems);
        ls.userAddedItem = true;
      });
      it('adds a new item', () => {
        let event = {
          nativeEvent: {
            text: 'bananas',
          },
        };
        ls.addNewItem(event);
        expect(ls.state.items).toEqual([
          {
            name: 'Apples',
            done: false,
          }, {
            name: 'Oranges',
            done: true,
          }, {
            name: 'Bananas',
            done: false,
          },
        ]);
        expect(ls.userAddedItem).toBe(true);
        expect(ls.listChanged).toBe(true);
      });

      it('doesn\'t add a new item if it\'s empty', () => {
        let event = {
          nativeEvent: {
            text: '',
          },
        };
        ls.addNewItem(event);
        let spy = jest.spyOn(ls, 'setState');
        expect(spy)
          .not
          .toBeCalled();
      });

      it('doesn\'t add a new item if it already exists', () => {
        let event = {
          nativeEvent: {
            text: 'oranges',
          },
        };
        ls.addNewItem(event);
        expect(ls.state.items).toEqual([
          {
            name: 'Apples',
            done: false,
          }, {
            name: 'Oranges',
            done: true,
            alreadyExistsFlag: true,
          },
        ]);
        expect(ls.spyItemAlreadyExists).toBeCalledWith('Oranges', 1);
      });

      it('places a new item over the placeholder item', () => {
        let event = {
          nativeEvent: {
            text: 'oranges',
          },
        };
        ls.state.items = [
          {
            name: 'placeholder',
            done: false,
          },
        ];
        ls.userAddedItem = false;
        ls.addNewItem(event);
        expect(ls.state.items).toEqual([
          {
            name: 'Oranges',
            done: false,
          },
        ]);
        expect(ls.userAddedItem).toBe(true);
      });
    });

    it('records an item to edit', () => {
      ls.state.items = _.cloneDeep(testItems);
      let item = testItems[1].name;
      ls.recordItemToEdit(item);
      expect(ls.state.itemToEdit).toEqual(testItems[1]);
      expect(mockPush).toBeCalledWith({title: 'Edit item', index: 1});
    });

    it('saves current list before exiting', () => {
      ls.state.items = _.cloneDeep(testItems);
      jest.mock('react-native');
      ls.saveList = jest.fn();
      ls.componentWillUnmount();
      expect(ls.saveList).toBeCalledWith('onExit');
      jest.unmock('react-native');
    });

    describe('editItem', () => {
      beforeEach(() => {
        ls.state.items = _.cloneDeep(testItems);
        ls.recordItemToEdit('Oranges');
      });

      it('edits the recorded item', () => {
        let event = {
          nativeEvent: {
            text: 'bananas',
          },
        };
        ls.editItem(event);
        expect(ls.state.items).toEqual([
          {
            name: 'Apples',
            done: false,
          }, {
            name: 'Bananas',
            done: true,
          },
        ]);
        expect(ls.listChanged).toBe(true);
        expect(ls.userAddedItem).toBe(true);
      });

      it('doesn\'t add an edited item if there is already' +
        ' an item with the same name',
      () => {
        let event = {
          nativeEvent: {
            text: 'apples',
          },
        };
        ls.editItem(event);
        expect(ls.state.items).toEqual([
          {
            'alreadyExistsFlag': true,
            'done': false,
            'name': 'Apples',
          }, {
            'done': true,
            'name': 'Oranges',
          },
        ]);
        expect(ls.spyItemAlreadyExists).toBeCalledWith('Apples', 0);
      });
    });

    it('generates the QR code and switched to QR scene', () => {
      let spySetState = jest.spyOn(ls, 'setState');
      qr.packItems.mockReturnValue(42);
      return ls.generateQR().then(() => {
        expect(ls.spyCloseMenuModal).toBeCalled();
        expect(spySetState.mock.calls[1][0]).toEqual({generatingQR: true});
        expect(spySetState.mock.calls[2][0]).toEqual({generatingQR: false});
        expect(RNFetchBlob.fs.files['/qr.png']).toBeDefined();
        expect(ls.props.navigator.push).toBeCalledWith({
          title: 'Show QRCode',
          index: ls.props.route.index + 1,
        });
      });
    });

    it('correctly sets the state after reading QR', () => {
      let testQRdata = [
        {name: 'Bananas', done: false},
        {name: 'Lemons', done: true}];
      qr.unpackItems.mockReturnValue(testQRdata);
      ls.processQR({type: 'QR_CODE'});
      expect(qr.unpackItems).toBeCalled();
      expect(ls.props.navigator.pop).toBeCalled();
      expect(ls.state.items).toEqual(testQRdata);
      expect(ls.listChanged).toBe(true);
    });

    it('scrolls to duplicate item instead of adding it', () => {
      ls.state.items = _.cloneDeep(testItems);
      ls.render();
      ls.componentDidMount();
      ls.itemAlreadyExists('Oranges', 1);
      expect(ls.state.items[1].name).toBe('Oranges');
      expect(ls.state.items[1].alreadyExistsFlag).toBe(true);
      expect(ls.itemList.listView.scrollTo.mock.calls[0][0]).toEqual({
        x: 0,
        y: -(ls.ds.getRowCount() * 45),
        animate: false,
      });
      expect(ls.itemList.listView.scrollTo.mock.calls[1][0]).toEqual({
        x: 0,
        y: 1 * 45,
        animate: false,
      });
    });

    describe('checkItem', () => {
      beforeEach(() => {
        ls.state.items = _.cloneDeep(testItems);
      });

      afterEach(() => {
        expect(ls.listChanged).toBe(true);
      });

      it('checks an unchecked item', () => {
        ls.checkItem('Apples');
        expect(ls.state.items[0].done).toBe(true);
      });

      it('unchecks a checked item', () => {
        ls.checkItem('Oranges');
        expect(ls.state.items[1].done).toBe(false);
      });
    });

    describe('newList', () => {
      it('clears the list', () => {
        ls.newList();
        expect(ls.state.items).toEqual([
          {name: ls.EMPTY_ITEM_TEXT, done: false}]);
        expect(ls.userAddedItem).toBe(false);
        expect(ls.listChanged).toBe(true);
      });
    });

    describe('Unsaved changes', () => {
      beforeEach(() => {
        Alert.alert = jest.fn();
        ls.listChanged = true;
        ls.userAddedItem = true;
      });

      it('warns before loading', () => {
        ls.loadListAlert();
        expect(ls.closeMenuModal).toBeCalled();
        let onPress = Alert.alert.mock.calls[0][2][0]['onPress'];
        expect(onPress.toString()).toBe(ls.loadList.bind(ls).toString());
        expect(Alert.alert).toBeCalledWith(i18n.t('loadListTitle'),
          i18n.t('loadListMsg'), [{text: 'OK', onPress: onPress},
          {text: 'Cancel'}]);
      });

      it('warns before creating a new list', () => {
        ls.newListAlert();
        expect(ls.closeMenuModal).toBeCalled();
        let onPress = Alert.alert.mock.calls[0][2][0]['onPress'];
        expect(onPress.toString()).toBe(ls.newList.bind(ls).toString());
        expect(Alert.alert).toBeCalledWith(i18n.t('createNewListTitle'),
          i18n.t('createNewListMsg'), [{text: 'OK', onPress: onPress},
          {text: 'Cancel'}]);
      });

      it('warns before switching to qr reading scene', () => {
        ls.readQRAlert();
        expect(ls.closeMenuModal).toBeCalled();
        let onPress = Alert.alert.mock.calls[0][2][0]['onPress'];
        expect(onPress.toString()).toBe(ls.readQR.bind(ls).toString());
        expect(Alert.alert).toBeCalledWith(i18n.t('readQRAlertTitle'),
          i18n.t('readQRAlertMsg'), [{text: 'OK', onPress: onPress},
          {text: 'Cancel'}]);
      });

      it('doesn\'t warn if there are no unsaved changes', () => {
        ls.listChanged = false;
        ls.userAddedItem = false;
        ls.loadList = jest.fn();
        ls.newList = jest.fn();
        ls.readQR = jest.fn();
        ls.closeMenuModal = jest.fn();
        ls.loadListAlert();
        ls.newListAlert();
        ls.readQRAlert();
        expect(ls.closeMenuModal.mock.calls.length).toBe(3);
        expect(Alert.alert).not.toBeCalled();
        expect(ls.loadList).toBeCalled();
        expect(ls.newList).toBeCalled();
        expect(ls.readQR).toBeCalled();
      });
    });

    describe('removeItem', () => {
      it('removes an item', () => {
        ls.state.items = _.cloneDeep(testItems);
        ls.removeItem('Oranges');
        expect(ls.state.items).toEqual([{
          name: 'Apples',
          done: false,
        }]);
        expect(ls.listChanged).toBe(true);
      });

      it('removes the last item and correctly inserts placeholder', () => {
        ls.state.items = [{name: 'Sausages', done: true}];
        ls.removeItem('Sausages');
        expect(ls.state.items).toEqual([{
          name: ls.EMPTY_ITEM_TEXT,
          done: false,
        }]);
        expect(ls.listChanged).toBe(true);
        expect(ls.userAddedItem).toBe(false);
      });
    });
  });
});
