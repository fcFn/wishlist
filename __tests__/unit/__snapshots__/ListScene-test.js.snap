// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListScene component renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#F5FCFF",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Modal
    animationType="none"
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    transparent={true}
    visible={false}
  >
    <View
      style={
        Object {
          "backgroundColor": "rgba(0,0,0,0.5)",
          "flex": 1,
        }
      }
    >
      <View
        style={
          Object {
            "alignSelf": "center",
            "backgroundColor": "white",
            "marginTop": 7,
            "width": 340,
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "fontWeight": "bold",
              "lineHeight": 15,
              "padding": 10,
              "paddingTop": 5,
            }
          }
        >
          Welcome back to Wishlist! Haven't seen you in a long time. Let me remind you how to use the app.
          
          
          To add an item, press the ⊕ button at the bottom of the list.
          
          
          To remove an item, swipe it to the left.
          
          
          To edit an item, press it and hold.
          
          
          To show additional options, press and hold the ⊕ button.
          
          
          Additional options:
          
          
          Generate QR: generates a QR Code from the current list.
          
          
          Read QR: reads a QR Code generated in Wishlist and creates a list out of it.
          
          
          Load list: loads a previously saved list.
          
          
          Save list: saves the current list, overwriting the previously saved list.
          
          
          New list: creates a new list.
          
          
          Help: shows this window.
        </Text>
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "#ffbcbc",
                "borderColor": "red",
                "borderWidth": 1,
                "height": 100,
                "justifyContent": "center",
                "margin": 10,
                "width": 300,
              },
              undefined,
            ]
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#333333",
                "margin": 20,
              }
            }
          >
            TouchableNativeFeedback is not supported on this platform!
          </Text>
        </View>
      </View>
    </View>
  </Modal>
  <Modal
    animationType="none"
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    transparent={true}
    visible={false}
  >
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      hitSlop={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "rgba(0,0,0,0.5)",
          "flex": 1,
          "justifyContent": "center",
        }
      }
      testID={undefined}
    >
      <View
        style={
          Object {
            "alignSelf": "center",
            "backgroundColor": "white",
            "marginTop": 10,
            "width": 300,
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "#ffbcbc",
                "borderColor": "red",
                "borderWidth": 1,
                "height": 100,
                "justifyContent": "center",
                "margin": 10,
                "width": 300,
              },
              undefined,
            ]
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#333333",
                "margin": 20,
              }
            }
          >
            TouchableNativeFeedback is not supported on this platform!
          </Text>
        </View>
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "#ffbcbc",
                "borderColor": "red",
                "borderWidth": 1,
                "height": 100,
                "justifyContent": "center",
                "margin": 10,
                "width": 300,
              },
              undefined,
            ]
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#333333",
                "margin": 20,
              }
            }
          >
            TouchableNativeFeedback is not supported on this platform!
          </Text>
        </View>
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "#ffbcbc",
                "borderColor": "red",
                "borderWidth": 1,
                "height": 100,
                "justifyContent": "center",
                "margin": 10,
                "width": 300,
              },
              undefined,
            ]
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#333333",
                "margin": 20,
              }
            }
          >
            TouchableNativeFeedback is not supported on this platform!
          </Text>
        </View>
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "#ffbcbc",
                "borderColor": "red",
                "borderWidth": 1,
                "height": 100,
                "justifyContent": "center",
                "margin": 10,
                "width": 300,
              },
              undefined,
            ]
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#333333",
                "margin": 20,
              }
            }
          >
            TouchableNativeFeedback is not supported on this platform!
          </Text>
        </View>
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "#ffbcbc",
                "borderColor": "red",
                "borderWidth": 1,
                "height": 100,
                "justifyContent": "center",
                "margin": 10,
                "width": 300,
              },
              undefined,
            ]
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#333333",
                "margin": 20,
              }
            }
          >
            TouchableNativeFeedback is not supported on this platform!
          </Text>
        </View>
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "#ffbcbc",
                "borderColor": "red",
                "borderWidth": 1,
                "height": 100,
                "justifyContent": "center",
                "margin": 10,
                "width": 300,
              },
              undefined,
            ]
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#333333",
                "margin": 20,
              }
            }
          >
            TouchableNativeFeedback is not supported on this platform!
          </Text>
        </View>
      </View>
    </View>
  </Modal>
  <RCTScrollView
    dataSource={
      ListViewDataSource {
        "items": 1,
      }
    }
    initialListSize={1}
    onMoveShouldSetResponder={[Function]}
    onMoveShouldSetResponderCapture={[Function]}
    onResponderEnd={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderReject={[Function]}
    onResponderRelease={[Function]}
    onResponderStart={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    onStartShouldSetResponderCapture={[Function]}
    removeClippedSubviews={false}
    renderRow={[Function]}
    renderScrollComponent={[Function]}
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View>
      <View
        onMoveShouldSetResponder={[Function]}
        onMoveShouldSetResponderCapture={[Function]}
        onResponderEnd={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderReject={[Function]}
        onResponderRelease={[Function]}
        onResponderStart={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        onStartShouldSetResponderCapture={[Function]}
        style={
          Object {
            "backgroundColor": "rgba(255, 255, 255, 1)",
            "borderLeftWidth": 0,
            "borderRightWidth": 0,
            "borderTopWidth": 0,
            "borderWidth": 2,
            "height": 40,
            "paddingLeft": 10,
            "transform": Array [
              Object {
                "translateX": 0,
              },
            ],
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "fontSize": 25,
            }
          }
        />
      </View>
    </View>
  </RCTScrollView>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#ffbcbc",
            "borderColor": "red",
            "borderWidth": 1,
            "height": 100,
            "justifyContent": "center",
            "margin": 10,
            "width": 300,
          },
          undefined,
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#333333",
            "margin": 20,
          }
        }
      >
        TouchableNativeFeedback is not supported on this platform!
      </Text>
    </View>
  </View>
</View>
`;
